commit a6d9daa2eeeb5b82c0b6fe1cb622fe815b733401
Author: linjiashuo <linjiashuo@wingtech.com>
Date:   Sat Mar 2 10:05:14 2019 +0800

    Issue Description:ExtB878208 shutdown device when cable is plug-out quickly after plugging in
    Root Cause:shutdown after running into charger service
    Product:K81923AA1
    Reference Git Repository:bootable/bootloader/edk2
    Modify Methods:shutdown before booting into kernel
    Test Requirements:redo cable-out test to see if shutdown quickly
    Owner:linjiashuo
    Reviewer:jinzhao
    Comments:shutdown device when cable is plug-out quickly after plugging in
    Range [1/2/3/4]:4
    
    Change-Id: Idc022909894a8f0d25ed09acc8b2f3d689572558
    (cherry picked from commit bb931981baa69fec90f117a6c3916e8a2af55cd1)

diff --git a/QcomModulePkg/Library/BootLib/BootLinux.c b/QcomModulePkg/Library/BootLib/BootLinux.c
index f3b4787..db2051e 100644
--- a/QcomModulePkg/Library/BootLib/BootLinux.c
+++ b/QcomModulePkg/Library/BootLib/BootLinux.c
@@ -49,6 +49,9 @@
 #include "libfdt.h"
 #include <ufdt_overlay.h>
 
+//ExtB878208, linjiashuo@wt, shutdown device when cable is plug-out quickly after plugging in, 20190301
+#include <Protocol/EFIChargerEx.h>
+
 STATIC QCOM_SCM_MODE_SWITCH_PROTOCOL *pQcomScmModeSwitchProtocol = NULL;
 STATIC BOOLEAN BootDevImage;
 STATIC BOOLEAN IsVmComputed = FALSE;
@@ -803,6 +806,12 @@ BootLinux (BootInfo *Info)
   HypMsg Msg = {0};
   UINT32 RetVal;
 
+  //+ ExtB878208, linjiashuo@wt, shutdown device when cable is plug-out quickly after plug-in, 20190301
+  EFI_CHARGER_EX_PROTOCOL *ChgDetectProtocol;
+  BOOLEAN IsOffModeCharging = FALSE;
+  BOOLEAN IsChargerPresence = FALSE;
+  //- ExtB878208, linjiashuo@wt, shutdown device when cable is plug-out quickly after plug-in, 20190301
+
   HypBootInfo *HypInfo = GetVmData ();
   if (IsVmEnabled () &&
       HypInfo == NULL) {
@@ -956,6 +965,38 @@ BootLinux (BootInfo *Info)
     return Status;
   }
 
+  //+ ExtB878208, linjiashuo@wt, shutdown device when cable is plug-out quickly after plugging in, 20190301
+  Status = gBS->LocateProtocol (&gChargerExProtocolGuid, NULL,
+                                (VOID **)&ChgDetectProtocol);
+  if (Status == EFI_NOT_FOUND) {
+    DEBUG ((EFI_D_VERBOSE, "Charger Protocol is not available.\n"));
+    return Status;
+  } else if (EFI_ERROR (Status)) {
+    DEBUG ((EFI_D_ERROR, "Error finding charger protocol: %r\n", Status));
+    return Status;
+  }
+
+  Status = ChgDetectProtocol->GetChargerPresence (&IsChargerPresence);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((EFI_D_ERROR, "Error getting IsChargerPresence info: %r\n", Status));
+    return Status;
+  }
+
+  Status = ChgDetectProtocol->IsOffModeCharging(&IsOffModeCharging);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((EFI_D_ERROR, "Error getting IsOffModeCharging info: %r\n", Status));
+    return Status;
+  }
+
+  DEBUG ((EFI_D_ERROR, "joshua: IsOffModeCharge=%d, IsChargerPresence=%d \n", IsOffModeCharging, IsChargerPresence));
+
+  if(IsOffModeCharging && !IsChargerPresence) {
+      DEBUG ((EFI_D_ERROR, "joshua: charging cable is plugged out, quit off-mode-charging, shutdown!!\n"));
+      ShutdownDevice();
+      return Status;
+  }
+  //- ExtB878208, linjiashuo@wt, shutdown device when cable is plug-out quickly after plugging in, 20190301
+
   Status = LoadAddrAndDTUpdate (&BootParamlistPtr);
   if (Status != EFI_SUCCESS) {
        return Status;
